transport-native-io_uring/src/main/c/netty_io_uring_native.c

```diff
@@ -0,0 +1,288 @@
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "syscall.h"
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <syscall.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include "netty_unix_buffer.h"
+#include "netty_unix_errors.h"
+#include "netty_unix_filedescriptor.h"
+#include "netty_unix_jni.h"
+#include "netty_unix_limits.h"
+#include "netty_unix_socket.h"
+#include "netty_unix_util.h"
+
+static jmethodID ringBufferMethodId = NULL;
+static jmethodID ioUringSubmissionQueueMethodId = NULL;
+static jmethodID ioUringCommpletionQueueMethodId = NULL;
+static jclass ringBufferClass = NULL;
+static jclass ioUringCompletionQueueClass = NULL;
+static jclass ioUringSubmissionQueueClass = NULL;
+
+void io_uring_unmap_rings(struct io_uring_sq *sq, struct io_uring_cq *cq) {
+  munmap(sq->ring_ptr, sq->ring_sz);
+  if (cq->ring_ptr && cq->ring_ptr != sq->ring_ptr)
+    munmap(cq->ring_ptr, cq->ring_sz);
+}
+
+int io_uring_mmap(int fd, struct io_uring_params *p, struct io_uring_sq *sq,
+                  struct io_uring_cq *cq) {
+  size_t size;
+  int ret;
+
+  sq->ring_sz = p->sq_off.array + p->sq_entries * sizeof(unsigned);
+  cq->ring_sz = p->cq_off.cqes + p->cq_entries * sizeof(struct io_uring_cqe);
+
+  if (p->features & IORING_FEAT_SINGLE_MMAP) {
+    if (cq->ring_sz > sq->ring_sz)
+      sq->ring_sz = cq->ring_sz;
+    cq->ring_sz = sq->ring_sz;
+  }
+  sq->ring_ptr = mmap(0, sq->ring_sz, PROT_READ | PROT_WRITE,
+                      MAP_SHARED | MAP_POPULATE, fd, IORING_OFF_SQ_RING);
+  if (sq->ring_ptr == MAP_FAILED)
+    return -errno;
+
+  if (p->features & IORING_FEAT_SINGLE_MMAP) {
+    cq->ring_ptr = sq->ring_ptr;
+  } else {
+    cq->ring_ptr = mmap(0, cq->ring_sz, PROT_READ | PROT_WRITE,
+                        MAP_SHARED | MAP_POPULATE, fd, IORING_OFF_CQ_RING);
+    if (cq->ring_ptr == MAP_FAILED) {
+      cq->ring_ptr = NULL;
+      ret = -errno;
+      goto err;
+    }
+  }
+
+  sq->khead = sq->ring_ptr + p->sq_off.head;
+  sq->ktail = sq->ring_ptr + p->sq_off.tail;
+  sq->kring_mask = sq->ring_ptr + p->sq_off.ring_mask;
+  sq->kring_entries = sq->ring_ptr + p->sq_off.ring_entries;
+  sq->kflags = sq->ring_ptr + p->sq_off.flags;
+  sq->kdropped = sq->ring_ptr + p->sq_off.dropped;
+  sq->array = sq->ring_ptr + p->sq_off.array;
+
+  size = p->sq_entries * sizeof(struct io_uring_sqe);
+  sq->sqes = mmap(0, size, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_POPULATE,
+                  fd, IORING_OFF_SQES);
+  if (sq->sqes == MAP_FAILED) {
+    ret = -errno;
+  err:
```

Comment:
@normanmaurer  yeah it would be easier to read, but we need boolean for no error because of unmap:
```
err:
     io_uring_unmap_rings(sq, cq);
     return ret;
```

Reply:
no boolean is required :relaxed:
```
	cq->khead = cq->ring_ptr + p->cq_off.head;
	cq->ktail = cq->ring_ptr + p->cq_off.tail;
	cq->kring_mask = cq->ring_ptr + p->cq_off.ring_mask;
	cq->kring_entries = cq->ring_ptr + p->cq_off.ring_entries;
	cq->koverflow = cq->ring_ptr + p->cq_off.overflow;
	cq->cqes = cq->ring_ptr + p->cq_off.cqes;

	return 0;
err:
	io_uring_unmap_rings(sq, cq);
	return ret;
```